File: demos/python/demo/celllist.py
# Copyright (c) 2015-2016 Lester Hedges <lester.hedges+vmmc@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

"""
@package demo
@author Lester Hedges
@brief Class for the cell list object.
"""

import math

class CellList:
    """ Class for the cell list. """

    def __init__(self, dimension = None, box_size = None, interaction_range = None):
        """ Cell list constructor. """

        # Set default dimension.
        self.dimension = 3

        if dimension is None and box_size is None and interaction_range is None:
            pass
        else:
            self.dimension = dimension
            self.initialise(box_size, interaction_range)

    def initialise(self, box_size, interaction_range):
        """ Initalise the cell list. """

        # Initialise data structures.

        self.cells_per_axis = [0] * self.dimension  # Number of cells per axis.
        self.cell_spacing = [0] * self.dimension    # Cell spacing along each axis.

        # Determine minimal cell spacing.
        for x in range(0, self.dimension):

            # Initialise number of cells.
            self.cells_per_axis[x] = 1

            while ((box_size[x] / self.cells_per_axis[x]) > interaction_range):
                self.cells_per_axis[x] += 1

            self.cells_per_axis[x] -=1
            self.cell_spacing[x] = box_size[x] / self.cells_per_axis[x]

            # Check that the number of cells per axis is large enough.
            if self.cells_per_axis[x] < 3:
                raise ValueError('Simulation box is too small for cell lists.')
                raise SystemExit

        # Estimate the number of particles per cell from the interaction range.
        # (Assumes the diameter is one.)
        if self.dimension is 3:
            self.max_particles = int((self.cell_spacing[0]*self.cell_spacing[1]*self.cell_spacing[2]) / ((4/3)*math.pi*0.5*0.5*0.5))
        else:
            self.max_particles = int((self.cell_spacing[0]*self.cell_spacing[1]) / (math.pi*0.5*0.5))

        # Add a buffer, e.g. if particles can overlap.
        self.max_particles += 10

        # Work out the total number of cells.
        self.num_cells = self.cells_per_axis[0]*self.cells_per_axis[1]
        if self.dimension is 3:
            self.num_cells *= self.cells_per_axis[2]

        # Resize data structures.

        self.index = [0] * self.num_cells       # Index of each cell.
        self.tally = [0] * self.num_cells       # Number of particles in each cell.
        self.particles = [0] * self.num_cells   # Indices of particles in each cell.
        self.neighbours = [0] * self.num_cells  # Indices of nearest neighbour cells.

        if self.dimension is 3:

            # Set number of nearest neighbour cells.
            self.num_neighbours = 27

            # Loop over all cells in x direction.
            for i in range(0, self.cells_per_axis[0]):

                # Loop over all cells in y direction.
                for j in range(0, self.cells_per_axis[1]):

                    # Loop over all cells in z direction.
                    for k in range(0, self.cells_per_axis[2]):

                        # Evaluate cell index.
                        m = i + self.cells_per_axis[0]*j + self.cells_per_axis[0]*self.cells_per_axis[1]*k

                        # Resize particle and nearest neighbour lists.
                        self.particles[m] = [0] * self.max_particles
                        self.neighbours[m] = [0] * self.num_neighbours

                        # Zero particle tally.
                        self.tally[m] = 0

                        # Set cell index
                        self.index[m] = m

                        # Initalise nearest neighbour counter
                        nn_count = 0

                        # x loop for nearest neighbours
                        for a in range(0, 3):

                            x = (i + (a-1) + self.cells_per_axis[0]) % self.cells_per_axis[0]

                            # y loop for nearest neighbours
                            for b in range(0, 3):

                                y = (j + (b-1) + self.cells_per_axis[1]) % self.cells_per_axis[1]

                                # z loop for nearest neighbours
                                for c in range(0, 3):

                                    z = (k + (c-1) + self.cells_per_axis[2]) % self.cells_per_axis[2]

                                    # Nearest neighbour cell index.
                                    nn = x + y*self.cells_per_axis[0] + z*self.cells_per_axis[0]*self.cells_per_axis[1]

                                    self.neighbours[m][nn_count] = nn
                                    nn_count += 1

        else:

            # Set number of nearest neighbour cells.
            self.num_neighbours = 9

            # Loop over all cells in x direction.
            for i in range(0, self.cells_per_axis[0]):

                # Loop over all cells in y direction.
                for j in range(0, self.cells_per_axis[1]):

                    # Evaluate cell index.
                    m = i + self.cells_per_axis[0]*j

                    # Resize particle and nearest neighbour lists.
                    self.particles[m] = [0] * self.num_neighbours
                    self.neighbours[m] = [0] * self.num_neighbours

                    # Zero particle tally.
                    self.tally[m] = 0

                    # Set cell index
                    self.index[m] = m

                    # Initalise nearest neighbour counter
                    nn_count = 0

                    # x loop for nearest neighbours
                    for a in range(0, 3):

                        x = (i + (a-1) + self.cells_per_axis[0]) % self.cells_per_axis[0]

                        # y loop for nearest neighbours
                        for b in range(0, 3):

                            y = (j + (b-1) + self.cells_per_axis[1]) % self.cells_per_axis[1]

                            # Nearest neighbour cell index.
                            nn = x + y*self.cells_per_axis[0]

                            self.neighbours[m][nn_count] = nn
                            nn_count += 1

    def reset(self):
        """ Reset cell tally counters. """
        self.tally = [0] * self.num_cells

    def get_cell(self, particle):
        """ Get a particle's cell index. """

        # Work out x and y indices.
        cellx = int(particle.position[0] / self.cell_spacing[0])
        celly = int(particle.position[1] / self.cell_spacing[1])

        # Compute the cell index.
        cell = cellx + celly * self.cells_per_axis[0]

        if self.dimension is 3:
            cellz = int(particle.position[2] / self.cell_spacing[2])
            cell += cellz * self.cells_per_axis[0] * self.cells_per_axis[1]

        return cell

    def init_cell(self, new_cell, particle):
        """ Insert a particle into the cell list. """

        # Add to new list.
        self.particles[new_cell][self.tally[new_cell]] = particle.index
        particle.cell = new_cell
        particle.pos_cell = self.tally[new_cell]
        self.tally[new_cell] += 1

        if self.tally[new_cell] == self.max_particles:
            raise ValueError('Maximum number of particles per cell exceeded.')
            raise SystemExit

    def init_cell_list(self, particles):
        """ Insert all particles into the cell list. """
        [init_cell(get_cell(particle), particle) for particle in particles]

    def update_cell(self, new_cell, particle, particles):
        """ Update the cell list. """

        # Remove particle from the old list.
        self.tally[particle.cell] -= 1
        self.particles[particle.cell][particle.pos_cell] = self.particles[particle.cell][self.tally[particle.cell]]
        particles[self.particles[particle.cell][self.tally[particle.cell]]].pos_cell = particle.pos_cell

        # Add particle to the new cell list
        self.particles[new_cell][self.tally[new_cell]] = particle.index
        particle.cell = new_cell
        particle.pos_cell = self.tally[new_cell]
        self.tally[new_cell] += 1

        if self.tally[new_cell] == self.max_particles:
            raise ValueError('Maximum number of particles per cell exceeded.')
            raise SystemExit

File: demos/src/CellList.cpp
/*
  Copyright (c) 2015-2016 Lester Hedges <lester.hedges+vmmc@gmail.com>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program. If not, see <http://www.gnu.org/licenses/>.
*/

#include <cmath>
#include <cstdlib>
#include <iostream>

#include "CellList.h"
#include "Particle.h"

#ifndef M_PI
    #define M_PI 3.1415926535897932384626433832795
#endif

CellList::CellList() : dimension(3)
{
}

CellList::CellList(unsigned int dimension_, const std::vector<double>& boxSize, double range) : dimension(dimension_)
{
    this->initialise(boxSize, range);
}

CellList& CellList::operator = (const CellList& cells)
{
    // resize cell list
    (*this).resize(cells.size());

    // copy across individual cells
    for (unsigned int i=0;i<cells.size();i++)
    {
        at(i) = cells[i];
    }

    return *this;
}

CellList& CellList::operator = (const std::vector<Cell>& cells)
{
    // resize cell list
    (*this).resize(cells.size());

    // copy across individual cells
    for (unsigned int i=0;i<cells.size();i++)
    {
        at(i) = cells[i];
    }

    return *this;
}

void CellList::initialise(const std::vector<double>& boxSize, double range)
{
    unsigned int i,j,k,m;
    unsigned int a,b,c;
    unsigned int x,y,z;
    unsigned int nn,nnCount;

    cellsPerAxis.resize(dimension);
    cellSpacing.resize(dimension);

    for (i=0;i<dimension;i++)
    {
        cellsPerAxis[i] = 1;

        while ((boxSize[i] / (double) cellsPerAxis[i]) > range)
        {
            cellsPerAxis[i]++;
        }
        cellsPerAxis[i]--;
        cellSpacing[i] = boxSize[i] / (double) cellsPerAxis[i];

        // check that number of cells per axis is large enough
        if (cellsPerAxis[i] < 3)
        {
            std::cerr << "[ERROR] CellList: Simulation box is too small (min cells per axis is 3)\n";
            exit(EXIT_FAILURE);
        }
    }

    // Estimate maximum number of particles per cell from interaction range.
    // (Assumes particle diameter is one.)
    if (dimension == 3) maxParticles = (cellSpacing[0]*cellSpacing[1]*cellSpacing[2]) / ((4.0/3.0)*M_PI*0.5*0.5*0.5);
    else maxParticles = (cellSpacing[0]*cellSpacing[1]) / (M_PI*0.5*0.5);

    // Add a buffer, e.g. if particles can overlap.
    maxParticles += 10;

    nCells = cellsPerAxis[0]*cellsPerAxis[1];
    if (dimension == 3) nCells *= cellsPerAxis[2];

    // resize cell list array
    (*this).resize(nCells);

    if (dimension == 3)
    {
        nNeighbours = 27;

        // loop over all cells x direction
        for (i=0;i<cellsPerAxis[0];i++)
        {
            // loop over all cells y direction
            for (j=0;j<cellsPerAxis[1];j++)
            {
                // loop over all cells z direction
                for (k=0;k<cellsPerAxis[2];k++)
                {
                    // cell index
                    m = i + cellsPerAxis[0]*j + cellsPerAxis[0]*cellsPerAxis[1]*k;

                    // resize neighbour list and particle arrays
                    at(m).neighbours.resize(nNeighbours);
                    at(m).particles.resize(maxParticles);

                    nnCount = 0;

                    // x loop for nearest neighbours
                    for (a=0;a<3;a++)
                    {
                        x = (i+(a-1)+cellsPerAxis[0])%cellsPerAxis[0];

                        // y loop for nearest neighbours
                        for (b=0;b<3;b++)
                        {
                            y = (j+(b-1)+cellsPerAxis[1])%cellsPerAxis[1];

                            // z loop for nearest neighbours
                            for (c=0;c<3;c++)
                            {
                                z = (k+(c-1)+cellsPerAxis[2])%cellsPerAxis[2];

                                // nn cell index
                                nn = x + y*cellsPerAxis[0] + z*cellsPerAxis[0]*cellsPerAxis[1];

                                at(m).neighbours[nnCount] = nn;
                                nnCount++;
                            }
                        }
                    }

                    // zero tally for each cell
                    at(m).tally = 0;
                    at(m).index = m;
                }
            }
        }
    }
    else
    {
        nNeighbours = 9;

        // loop over all cells x direction
        for (i=0;i<cellsPerAxis[0];i++)
        {
            // loop over all cells y direction
            for (j=0;j<cellsPerAxis[1];j++)
            {
                // cell index
                m = i + cellsPerAxis[0]*j;

                // resize neighbour list and particle arrays
                at(m).neighbours.resize(nNeighbours);
                at(m).particles.resize(maxParticles);

                nnCount = 0;

                // x loop for nearest neighbours
                for (a=0;a<3;a++)
                {
                    x = (i+(a-1)+cellsPerAxis[0])%cellsPerAxis[0];

                    // y loop for nearest neighbours
                    for (b=0;b<3;b++)
                    {
                        y = (j+(b-1)+cellsPerAxis[1])%cellsPerAxis[1];

                        // nn cell index
                        nn = x + y*cellsPerAxis[0];

                        at(m).neighbours[nnCount] = nn;
                        nnCount++;
                    }
                }

                // zero tally for each cell
                at(m).tally = 0;
                at(m).index = m;
            }
        }
    }
}

void CellList::reset()
{
    for (unsigned int i=0;i<nCells;i++) at(i).tally = 0;
}

int CellList::getCell(const Particle& particle)
{
    int cell,cellx,celly;

    cellx = int(particle.position[0]/cellSpacing[0]);
    celly = int(particle.position[1]/cellSpacing[1]);

    cell = cellx + celly*cellsPerAxis[0];

    if (dimension == 3)
    {
        int cellz = int(particle.position[2]/cellSpacing[2]);
        cell += cellz*cellsPerAxis[0]*cellsPerAxis[1];
    }

    return cell;
}

void CellList::initCell(int newCell, Particle& particle)
{
    // Add to new list
    at(newCell).particles[at(newCell).tally] = particle.index;
    particle.cell = newCell;
    particle.posCell = at(newCell).tally;
    at(newCell).tally++;

    if (at(newCell).tally == maxParticles)
    {
        std::cerr << "[ERROR] CellList: Maximum number of particles per cell exceeded!\n";
        exit(EXIT_FAILURE);
    }
}

void CellList::initCellList(std::vector<Particle>& particles)
{
    for (unsigned int i=0;i<particles.size();i++)
    {
        initCell(getCell(particles[i]), particles[i]);
    }
}

void CellList::updateCell(int newCell, Particle& particle, std::vector<Particle>& particles)
{
    // Remove from old list
    at(particle.cell).tally--;
    at(particle.cell).particles[particle.posCell] = at(particle.cell).particles[at(particle.cell).tally];
    particles[at(particle.cell).particles[at(particle.cell).tally]].posCell = particle.posCell;

    // Add to new list
    at(newCell).particles[at(newCell).tally] = particle.index;
    particle.cell = newCell;
    particle.posCell = at(newCell).tally;
    at(newCell).tally++;

    if (at(newCell).tally == maxParticles)
    {
        std::cerr << "[ERROR] CellList: Maximum number of particles per cell exceeded!\n";
        exit(EXIT_FAILURE);
    }
}

void CellList::setDimension(unsigned int dimension_)
{
    dimension = dimension_;
}

unsigned int CellList::getNeighbours() const
{
    return nNeighbours;
}
